Forecasting and anomaly detection

CampfireRemnants opened the pull request:
## What does this PR change?
Adds deployment information for forecasting and anomaly detection. For the two env vars it defines, I also want to make sure my understanding that they can be overwritten by a values.yaml is correct, because right now they're defaulted to the same dev values used prior, and we'll want to change that. Also KCM_BASE_URL probably should get a different default.

## Does this PR rely on any other PRs?
Doesn't rely on #2890 , but is an extension of it.

## How does this PR impact users? (This is the kind of thing that goes in release notes!)
Allows enterprise users to enable forecasting and anomaly detection!

## Links to Issues or tickets this PR addresses or fixes

<!--
Please use GithHub's closing keywords to link to any issue(s) this PR addresses. See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue how to use closing keywords.
-->



## What risks are associated with merging this PR? What is required to fully test this PR?
It shouldn't affect any other important KCM components, so the main risk is the embarrassment of non-functional forecasting/anomaly detection if those break.

## How was this PR tested?
In one of the QA clusters.

## Have you made an update to documentation? If so, please provide the corresponding PR.
Nope, still need to work with Brett on that

JesseGoodier committed:
forecasting init

JesseGoodier committed:
forecasting init

JesseGoodier committed:
forecasting init

JesseGoodier committed:
forecasting init

JesseGoodier committed:
forecasting init

JesseGoodier committed:
nginx edit

JesseGoodier committed:
nginx path

Commit:
Added two required env vars for forecasting.

thomasvn commented on cost-analyzer/templates/cost-analyzer-frontend-config-map-template.yaml:
```diff
@@ -102,6 +102,19 @@ data:
 {{- end }}
 
   {{- if and (not .Values.agent) (not .Values.cloudAgent) (not (eq (include "aggregator.deployMethod" .) "disabled")) }}
+  {{- if and (.Values.kubecostDeployment) (.Values.kubecostDeployment.queryServiceReplicas) (gt (.Values.kubecostDeployment.queryServiceReplicas | toString | atoi) 0) }}
+    upstream queryservice {
+        server {{ .Release.Name }}-query-service-load-balancer.{{ .Release.Namespace }}:9003;
+    }
+  {{- end }}
```
The diff shows that you're adding in some lines here referencing `queryServiceReplicas`. Was that intentional? My understanding was that QSR is no more?

CampfireRemnants replied:
Nope, unintentional, that was a mistake when I was resolving merge conflicts, thanks for the catch

michaelmdresser replied:
Addressed

thomasvn commented on cost-analyzer/templates/forecasting-deployment.yaml:
```diff
@@ -0,0 +1,102 @@
+            - name: CONFIG_PATH
+              value: /var/configs/
+            - name: KCM_BASE_URL
+              value: "https://aggregator.nightly.kubecost.xyz/model"
```
```suggestion
              value: {{ template "cost-analyzer.serviceName" . }}
```

This _should_ work. This will point to the k8s DNS name of the `cost-analyzer` service. You may need to add `:9003/model` as well.

michaelmdresser replied:
Yes, this is the right thing to do. As Thomas said, you may need to add a port and route -- _not_ 9003 though! `:9090/model`, or `:9004` if you want to hit aggregator directly. 9003 is the cost-model container.

thomasvn replied:
Oops! Great catch thank you

thomasvn commented on cost-analyzer/templates/forecasting-deployment.yaml:
```diff
@@ -0,0 +1,102 @@
+              value: /var/configs/
+            - name: KCM_BASE_URL
+              value: "https://aggregator.nightly.kubecost.xyz/model"
+            - name: MODEL_STORAGE_PATH
+              value: "/tmp/localrun/models"
```
I assume you're writing your models to disk at this location? How large are these models? If they're sufficiently large we may want to consider creating a PersistentVolume for them.

michaelmdresser replied:
They are starting out relatively simple, and we're focusing on functionality first over perfect stability of saved models, so this isn't a major initial concern. This is an excellent general callout, though! We have to follow up here when the models get more sophisticated and the component has been proven out more.

chipzoller replied:
Ditto. Also note because this does not yet adopt global securityContext, writing to `/tmp` may be disallowed with `readOnlyRootFilesystem` set to `true`.

michaelmdresser replied:
With r/o root fs, I added an empty dir volume to make this work correctly. In the future we'll want to consider the PV question more closely and ideally make the path configurable.

chipzoller replied:
That works, but there should probably be a size limit set for this. emptyDir's default backing medium is memory. Do we know how large these models are, per Thomas' question?

CampfireRemnants replied:
Currently, the models are very small json files, generally 1-2 Mb for all of them

michaelmdresser replied:
Thanks Chip, I didn't know that -- set sizeLimit on the emptydir to 500Mi to start

thomasvn commented on cost-analyzer/templates/forecasting-service-account.yaml:
```diff
@@ -0,0 +1,15 @@
+{{- if .Values.forecasting.enabled }}
+{{- if and .Values.serviceAccount.create .Values.forecasting.serviceAccountName }}
+apiVersion: v1
+kind: ServiceAccount
```
I don't think we need this file? @jessegoodier Any reason you created it in the initial draft?

michaelmdresser replied:
Good point Thomas, I removed this. If it causes problems it can be reintroduced.

thomasvn commented on cost-analyzer/values.yaml:
```diff
@@ -2235,6 +2235,52 @@ kubecostDeployment:
   labels: {}
   annotations: {}
 
+  ## QueryServiceReplicas
```
Again, not sure if you meant to add back in QueryServiceReplicas here?

michaelmdresser replied:
Removed, nice catch.

thomasvn commented:
@CampfireRemnants Looking really good! I'm noticing this CI failure though:

```
Error:  templates/: parse error at (cost-analyzer/templates/cost-analyzer-frontend-config-map-template.yaml:940): unexpected EOF
```

As you make changes to this draft, it may help to continuously run `helm template ./cost-analyzer --set forecasting.enabled=true` to make sure that helm is templating all these variables as you expect it to.

chipzoller reviewed (changes requested):
Let's please not start another new root-level values tree here. Let's start moving towards some semblance of organizing these values into logical structures. Forecasting is a product configuration so how about we start something like an `appConfig` at the root instead?

Also, please ensure all new values are defined (even if commented out) in the main values file. There shouldn't be any new values unaccounted for.

chipzoller commented:
I'm not opposed to putting more things at root level but I think there's a better organizational strategy here. Perhaps some other abstraction layer?

nealormsbee committed:
Add location block for anomaly detection

chipzoller committed:
Merge branch 'develop' into ashleigh/forecasting-and-anomaly-detection

michaelmdresser committed:
Address forecasting PR feedback (#2947)

* Fix upstream definitions

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

* Use K8s service for KCM_BASE_URL

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

* Remove forecasting service account

If it is needed, it can be resurrected

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

* Remove QSR values

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

---------

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser committed:
Fix URL typo

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser committed:
Remove unnecessary openSourceOnly check

Handled with a fail in helpers.tpl already

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser committed:
Remove erroneous {{end}}

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

chipzoller commented on cost-analyzer/values.yaml:
```diff
@@ -2235,6 +2235,10 @@ kubecostDeployment:
   labels: {}
   annotations: {}
 
+## Kubecost Forecasting
```
Please provide more description for what this feature is and what it does.

michaelmdresser replied:
Done!

michaelmdresser committed:
Remove incorrect image templating logic

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser committed:
Document all forecasting values in values.yaml

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser committed:
Explain Kubecost Forecasting in values.yaml

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

chipzoller commented on cost-analyzer/templates/forecasting-deployment.yaml:
Needs to incorporate the following common elements from the chart:

* Global Pod annotations
* Global labels
* Global securityContext and containerSecurityContext
* Fixed probes and, preferably, templatization similar to cost-model and frontend

michaelmdresser replied:
First three should be done now. The probes are an in-code requirement and out of my control on this PR at the moment.

What do you mean by "templatization similar to cost-model and frontend"?

chipzoller replied:
Liveness and readiness probes are templated for cost-model and frontend, per a more recent change. We should strive for consistency to do the same across all our other Pod controllers.

chipzoller replied:
Ref: https://github.com/kubecost/cost-analyzer-helm-chart/blob/develop/cost-analyzer/values.yaml#L549-L560

michaelmdresser replied:
Got it, thank you. I don't believe Ashleigh has the necessary endpoints implemented on this container yet.

CampfireRemnants replied:
@chipzoller  For the liveness and readiness probes, any endpoints other than just `/healthz` required?

michaelmdresser replied:
Chart updated for the upcoming /healthz -- we will use that for both liveness and readiness Ashleigh, so we just need a 200 from that.

chipzoller replied:
`/healthz` is pretty standard, but the health check should really not have any external dependencies in order to return a successful code. We don't do that for other containers, but this is one of the best practices.

chipzoller replied:
Resolving this conversation since the current state of this PR does not have probes. My understanding from this discussion is that probes will be implemented in a follow-up PR once the code is incorporated and a new image has been published. If that's not right, please correct me.

chipzoller commented on cost-analyzer/templates/forecasting-service.yaml:
```diff
@@ -0,0 +1,18 @@
+spec:
+  selector:
+{{ include "forecasting.selectorLabels" . | nindent 4 }}
+  type: "ClusterIP"
```
```suggestion
  type:ClusterI
```

chipzoller commented on cost-analyzer/templates/forecasting-service.yaml:
```diff
@@ -0,0 +1,18 @@
+  name: {{ template "forecasting.serviceName" . }}
+  namespace: {{ .Release.Namespace }}
+  labels:
+{{ include "forecasting.commonLabels" . | nindent 4 }}
```
```suggestion
    {{ include "forecasting.commonLabels" . | nindent 4 }}
```

michaelmdresser replied:
This is correctly indented when I open the file in my local editor.

michaelmdresser replied:
Juust kidding, different file, will fix

michaelmdresser committed:
Single proxy rule for /forecasting to forecasting

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

chipzoller commented on cost-analyzer/templates/forecasting-service.yaml:
```diff
@@ -0,0 +1,18 @@
+{{ include "forecasting.commonLabels" . | nindent 4 }}
+spec:
+  selector:
+{{ include "forecasting.selectorLabels" . | nindent 4 }}
```
```suggestion
    {{ include "forecasting.selectorLabels" . | nindent 4 }}
```

michaelmdresser replied:
This is correctly indented when I open the file in my local editor.

michaelmdresser replied:
Fixed

michaelmdresser committed:
Add global labels

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser committed:
Remove quotes

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser committed:
Fix template indendation

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

chipzoller reviewed (commented):
This needs to be tested with global level securityContext entries populated for both Pod- and container-level securityContext settings. All Deployments/Pods need to be able to pass the Pod Security Standards (PSS) restricted profile.

michaelmdresser committed:
Add security contexts and necessary emptyDir vol

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

chipzoller commented on cost-analyzer/templates/forecasting-deployment.yaml:
```diff
@@ -0,0 +1,124 @@
+          image: gcr.io/kubecost1/forecasting:prod-{{ $.Chart.AppVersion }}
+          {{ end }}
+          args: ["forecasting"]
+          {{/*
+          fix readinessProbe
+
+          readinessProbe:
+            httpGet:
+              path: /healthz
+              port: 5000
+            initialDelaySeconds: 10
+            periodSeconds: 5
+            failureThreshold: 200
+          */}}
```
Commented out. Need to fix.

michaelmdresser replied:
@CampfireRemnants can you make flask serve `/healthz` and always return a 200? That's a prerequisite

CampfireRemnants replied:
Yes, easily.

michaelmdresser replied:
Probes fixed. Will test once an image with /healthz is available.

michaelmdresser committed:
Fix and explain tmp volume mount

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser commented:
Security contexts have been added and this chart is now installed in a live cluster; the container appears to be working correctly.

michaelmdresser committed:
Set initial image for forecasting

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser committed:
liveness and readiness probes for forecasting

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser committed:
Set emptyDir size limit for forecasting

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser commented:
@chipzoller I think I've responded to all of the feedback on Ashleigh's PR here. I know we've talked about moving values out of the root in the past; I agree with the general idea but I don't feel like I have a good concept of the structure we should be pursuing, so I haven't made a change there.

chipzoller commented:
Is this ready for final review or there still outstanding items?

chipzoller commented on cost-analyzer/values.yaml:
```diff
@@ -2235,6 +2235,46 @@ kubecostDeployment:
+  fullImageName: gcr.io/kubecost1/kubecost-modeling:6ff46def0575545eb37c5dae179e5035a1de279f
+
+  # Resource specification block for the forecasting container.
+  resources: {}
```
Are there truly no resources we can offer as defaults here?

michaelmdresser replied:
A good point. I took a stab at some, let me know if they seem reasonable

chipzoller commented on cost-analyzer/values.yaml:
```diff
@@ -2235,6 +2235,46 @@ kubecostDeployment:
+  # image provided (registry, image, tag) will be used for the forecasting
+  # container.
+  # Example: fullImageName: gcr.io/kubecost1/forecasting:v0.0.1
+  fullImageName: gcr.io/kubecost1/kubecost-modeling:6ff46def0575545eb37c5dae179e5035a1de279f
```
What is this tag?

michaelmdresser replied:
We plan to release forecasting like Cluster Controller and Network Costs, with a fixed image in the Helm chart (instead of the branching/tagging used for higher-traffic repos like KCM). The particular images being set here are development images that are very close to the final release state.

chipzoller commented on cost-analyzer/templates/forecasting-deployment.yaml:
```diff
@@ -0,0 +1,134 @@
+    matchLabels:
+      {{ include "forecasting.selectorLabels" . | nindent 6 }}
+  strategy:
+    type: Recreate
```
cost-analyzer defaults to `RollingUpdate`. Are we sure we want `Recreate` here?

michaelmdresser replied:
Switched to RollingUpdate for consistency with other Deployments, I'm not sure why Recreate was chosen initially.

chipzoller reviewed (commented):
See comments thus far. Please also confirm this Deployment does not need a ServiceAccount of any sort. If it does not, we should disable the automounting of one (`spec.template.spec.automountServiceAccountToken: false`).

chipzoller committed:
optimize, clean

Signed-off-by: GitHub <noreply@github.com>

chipzoller committed:
remove newline

Signed-off-by: GitHub <noreply@github.com>

michaelmdresser committed:
Remove unnecessary args

The gunicorn switch means this isn't necessary any more

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser committed:
Add default resources block for forecasting

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser committed:
Disable SA token mount because no SA is needed

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser committed:
Update forecasting image

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser committed:
Switch to RollingUpdate for consistency

Other Deployments use RollingUpdate, there is no particular reason
to use Recreate

Signed-off-by: Michael Dresser <michaelmdresser@gmail.com>

michaelmdresser commented:
Responded to all -- AFAIK we don't need a service account (runs fine without) so that is gone and I set the value you suggested be set in that case.
